vmupdate/vmupdate.py

```diff
@@ -90,17 +111,124 @@ def get_targets(args, app):
+            try:
+                to_update = vm.features.get('updates-available', False)
+            except qubesadmin.exc.QubesDaemonCommunicationError:
+                to_update = False
+
+            if not to_update:
+                to_update = stale_update_info(vm, args)
+
+            if to_update:
+                targets.append(vm)
+
+    return targets
+
+
+def stale_update_info(vm, args):
+    today = datetime.today()
+    try:
+        last_update_str = vm.features.get(
+            'last-updates-check',
+            datetime.fromtimestamp(0).strftime('%Y-%m-%d %H:%M:%S')
+        )
+        last_update = datetime.fromisoformat(last_update_str)
+        if (today - last_update).days > args.n:
+            return True
+    except qubesadmin.exc.QubesDaemonCommunicationError:
+        pass
+    return False
+
+def run_update(targets, args, qube_klass="qubes"):
+    if args.dry_run:
+        print(f"Following {qube_klass} will be updated:",
+              ",".join((target.name for target in targets)))
+        return 0
+
+    runner = update_manager.UpdateManager(targets, args)
     return runner.run(agent_args=args)
 
 
+def restart_app_vms(args, templates):
+    if not args.restart:
+        return
+
+    templates_to_shutdown = [template for template in templates
+                             if template.is_running()]
+    if args.dry_run:
+        print("Following templates will be shutdown:",
+              ",".join((target.name for target in templates_to_shutdown)))
+        # we do not check if any volume is outdated, we expect it will be.
+        to_restart = {vm
+                      for template in templates
+                      for vm in template.appvms
+                      if vm.klass in ('AppVM', 'DispVM')
+                      and vm.is_running()}
+        print("Following qubes CAN be restarted:",
+              ",".join((target.name for target in to_restart)))
+        return
+
+    # first shutdown templates to apply changes to the root volume
+    for template in templates_to_shutdown:
+        try:
+            template.shutdown()
+        except qubesadmin.exc.QubesVMError as exc:
+            print(exc, file=sys.stderr)
+
+    to_restart = {vm
+                  for template in templates
+                  for vm in template.appvms
+                  if vm.klass in ('AppVM', 'DispVM')
+                  and vm.is_running()
+                  and any(vol.is_outdated() for vol in vm.volumes.values())}
+
+    restart_vms(to_restart)
+
+    # they are no need to start templates automatically
+
+
+def restart_vms(to_restart):
+    # try shutdown/kill qubes
+    wait_for = []
+    for vm in to_restart:
+        if next(vm.connected_vms(), None) is None:
+            try:
+                vm.shutdown()
+                wait_for.append(vm)
+            except qubesadmin.exc.QubesVMError as exc:
+                print(exc, file=sys.stderr)
+        else:
+            try:
+                vm.kill()
```

Comment:
should be `vm.shutdown(force=True)` - it does graceful shutdown, but ignores dependencies check (like networking). In fact, you can always do `vm.shutdown(force=True)` (as you're going to restart it back) and don't need to check connected_vms()